From a6c2a7e38929d965984228f72bb0b933d116be2f Mon Sep 17 00:00:00 2001
From: Luciana Fujii Pontello <luciana.fujii@collabora.co.uk>
Date: Fri, 17 Jun 2011 16:50:38 -0300
Subject: [PATCH 13/19] effectv: enable renegotiation flushing mid-way

Re-enable re-negotiation for shagadelic, vertigo and warp effects.
It flushes when the re-negotiation happens to make the new media be
played as soon as possible.
---
 gst/effectv/gstshagadelic.c |   13 ++++++++++---
 gst/effectv/gstvertigo.c    |   14 +++++++++++---
 gst/effectv/gstwarp.c       |   13 +++++++++++--
 3 files changed, 32 insertions(+), 8 deletions(-)

diff --git a/gst/effectv/gstshagadelic.c b/gst/effectv/gstshagadelic.c
index 650a563..ee2c207 100644
--- a/gst/effectv/gstshagadelic.c
+++ b/gst/effectv/gstshagadelic.c
@@ -83,9 +83,17 @@ gst_shagadelictv_set_caps (GstBaseTransform * btrans, GstCaps * incaps,
   GstShagadelicTV *filter = GST_SHAGADELICTV (btrans);
   GstStructure *structure;
   gboolean ret = FALSE;
+  gboolean flush = FALSE;
+  GstPad *pad = GST_BASE_TRANSFORM_SRC_PAD (btrans);
 
   structure = gst_caps_get_structure (incaps, 0);
 
+  /* If it is renegotiating mid-way we should flush */
+  if (filter->width != 0) {
+    gst_pad_push_event (pad, gst_event_new_flush_start ());
+    flush = TRUE;
+  }
+
   GST_OBJECT_LOCK (filter);
   if (gst_structure_get_int (structure, "width", &filter->width) &&
       gst_structure_get_int (structure, "height", &filter->height)) {
@@ -102,6 +110,8 @@ gst_shagadelictv_set_caps (GstBaseTransform * btrans, GstCaps * incaps,
   }
   GST_OBJECT_UNLOCK (filter);
 
+  if (flush)
+    gst_pad_push_event (pad, gst_event_new_flush_stop ());
   return ret;
 }
 
@@ -267,7 +277,4 @@ gst_shagadelictv_init (GstShagadelicTV * filter, GstShagadelicTVClass * klass)
 {
   filter->ripple = NULL;
   filter->spiral = NULL;
-
-  gst_pad_use_fixed_caps (GST_BASE_TRANSFORM_SRC_PAD (filter));
-  gst_pad_use_fixed_caps (GST_BASE_TRANSFORM_SINK_PAD (filter));
 }
diff --git a/gst/effectv/gstvertigo.c b/gst/effectv/gstvertigo.c
index 2f7ac7c..adeac08 100644
--- a/gst/effectv/gstvertigo.c
+++ b/gst/effectv/gstvertigo.c
@@ -84,9 +84,17 @@ gst_vertigotv_set_caps (GstBaseTransform * btrans, GstCaps * incaps,
   GstVertigoTV *filter = GST_VERTIGOTV (btrans);
   GstStructure *structure;
   gboolean ret = FALSE;
+  gboolean flush = FALSE;
+  GstPad *pad = GST_BASE_TRANSFORM_SRC_PAD (btrans);
 
   structure = gst_caps_get_structure (incaps, 0);
 
+  /* If it is renegotiating mid-way we should flush */
+  if (filter->width != 0) {
+    gst_pad_push_event (pad, gst_event_new_flush_start ());
+    flush = TRUE;
+  }
+
   GST_OBJECT_LOCK (filter);
   if (gst_structure_get_int (structure, "width", &filter->width) &&
       gst_structure_get_int (structure, "height", &filter->height)) {
@@ -103,6 +111,8 @@ gst_vertigotv_set_caps (GstBaseTransform * btrans, GstCaps * incaps,
   }
   GST_OBJECT_UNLOCK (filter);
 
+  if (flush)
+    gst_pad_push_event (pad, gst_event_new_flush_stop ());
   return ret;
 }
 
@@ -325,7 +335,5 @@ gst_vertigotv_init (GstVertigoTV * filter, GstVertigoTVClass * klass)
   filter->phase = 0.0;
   filter->phase_increment = 0.02;
   filter->zoomrate = 1.01;
-
-  gst_pad_use_fixed_caps (GST_BASE_TRANSFORM_SRC_PAD (filter));
-  gst_pad_use_fixed_caps (GST_BASE_TRANSFORM_SINK_PAD (filter));
+  filter->width = 0;
 }
diff --git a/gst/effectv/gstwarp.c b/gst/effectv/gstwarp.c
index e5eb758..2fef975 100644
--- a/gst/effectv/gstwarp.c
+++ b/gst/effectv/gstwarp.c
@@ -90,9 +90,17 @@ gst_warptv_set_caps (GstBaseTransform * btrans, GstCaps * incaps,
   GstWarpTV *filter = GST_WARPTV (btrans);
   GstStructure *structure;
   gboolean ret = FALSE;
+  gboolean flush = FALSE;
+  GstPad *pad = GST_BASE_TRANSFORM_SRC_PAD (btrans);
 
   structure = gst_caps_get_structure (incaps, 0);
 
+  /* If it is renegotiating mid-way we should flush */
+  if (filter->width != 0) {
+    gst_pad_push_event (pad, gst_event_new_flush_start ());
+    flush = TRUE;
+  }
+
   GST_OBJECT_LOCK (filter);
   if (gst_structure_get_int (structure, "width", &filter->width) &&
       gst_structure_get_int (structure, "height", &filter->height)) {
@@ -109,6 +117,9 @@ gst_warptv_set_caps (GstBaseTransform * btrans, GstCaps * incaps,
   }
   GST_OBJECT_UNLOCK (filter);
 
+  if (flush)
+    gst_pad_push_event (pad, gst_event_new_flush_stop ());
+
   return ret;
 }
 
@@ -288,6 +299,4 @@ gst_warptv_class_init (GstWarpTVClass * klass)
 static void
 gst_warptv_init (GstWarpTV * warptv, GstWarpTVClass * klass)
 {
-  gst_pad_use_fixed_caps (GST_BASE_TRANSFORM_SRC_PAD (warptv));
-  gst_pad_use_fixed_caps (GST_BASE_TRANSFORM_SINK_PAD (warptv));
 }
-- 
1.7.1

